ingestion/src/test/java/feast/ingestion/ImportJobTest.java

```diff
@@ -180,91 +200,376 @@ public void runPipeline_ShouldWriteToRedisCorrectlyGivenValidSpecAndFeatureRow()
+      throws IOException, InterruptedException {
+
+    List<FeatureRow> input = new ArrayList<>();
+
+    IntStream.range(0, IMPORT_JOB_SAMPLE_FEATURE_ROW_SIZE)
+        .forEach(
+            i -> {
+              FeatureRow row = defaultValues.get(i);
+              row =
+                  row.toBuilder()
+                      .clone()
+                      .addFields(
+                          Field.newBuilder()
+                              .setName("feature_4")
+                              .setValue(Value.newBuilder().setInt32Val(0).build()))
                       .build();
-              expected.put(redisKey, randomRow);
+              input.add(row);
             });
 
-    LOGGER.info("Starting Import Job with the following options: {}", options.toString());
-    PipelineResult pipelineResult = ImportJob.runPipeline(options);
-    Thread.sleep(Duration.standardSeconds(IMPORT_JOB_READY_DURATION_SEC).getMillis());
+    Map<RedisKey, FeatureRow> expected = getExpectedRedisMapping(input, spec);
+
+    PipelineResult pipelineResult = startImportJob(options);
     Assert.assertEquals(pipelineResult.getState(), State.RUNNING);
+    publishFeatureRows(input, pipelineResult);
 
-    LOGGER.info("Publishing {} Feature Row messages to Kafka ...", input.size());
-    TestUtil.publishFeatureRowsToKafka(
-        KAFKA_BOOTSTRAP_SERVERS,
-        KAFKA_TOPIC,
-        input,
-        ByteArraySerializer.class,
-        KAFKA_PUBLISH_TIMEOUT_SEC);
-    TestUtil.waitUntilAllElementsAreWrittenToStore(
-        pipelineResult,
-        Duration.standardSeconds(IMPORT_JOB_MAX_RUN_DURATION_SEC),
-        Duration.standardSeconds(IMPORT_JOB_CHECK_INTERVAL_DURATION_SEC));
+    LOGGER.info("Validating the actual values written to Redis ...");
+    StatefulRedisConnection<byte[], byte[]> connection = redisClient.connect(new ByteArrayCodec());
+    RedisCommands<byte[], byte[]> sync = connection.sync();
+    expected.forEach(
+        (key, expectedValue) -> {
+
+          // Ensure ingested key exists.
+          byte[] actualByteValue = getBytesFromKey(sync, key);
+          if (actualByteValue == null) {
+            Assert.fail("Missing key in Redis.");
+          }
+
+          // Ensure value is a valid serialized FeatureRow object.
+          FeatureRow actualValue = getFeatureRowFromByte(key, actualByteValue);
+
+          // Ensure the retrieved FeatureRow is equal to the ingested FeatureRow.
+          Assert.assertEquals(expectedValue, actualValue);
+        });
+    redisClient.shutdown();
+  }
+
+  /*
+   If there are missing fields in an incoming feature row, this test verifies that their values are set to null during ingestion.
+  */
+
+  @Test
+  public void runPipeline_ShouldWriteToRedisCorrectlyGivenMissingFieldsInFeatureRow()
+      throws IOException, InterruptedException {
+
+    List<FeatureRow> input = new ArrayList<>();
+
+    LOGGER.info("Generating test data ...");
+    IntStream.range(0, IMPORT_JOB_SAMPLE_FEATURE_ROW_SIZE)
```

Comment:
Can we just use a single FeatureRow here that is constructed by hand? That would make it a lot easier to understand what is going on. I think we should be testing missing fields inbetween other fields. 

Imagine a good row
TS E1 E2 F1 F2 F3 F4

Take away one feature
TS E1 E2 F1 __ F3 F4

The row should be stored as is above, and should be readable